File: go/server/turn_cancel_test.go
package server

import (
	"context"
	"encoding/json"
	"sync/atomic"
	"testing"
	"time"

	core "github.com/SmooAI/smooth-operator-core/go/core"
	"github.com/SmooAI/smooth-operator/go/protocol"
)

// User-initiated turn cancellation — the `cancel` action ("Stop button"). The Go port of
// the reference rust/smooth-operator-server/tests/turn_cancel.rs, driven over a real
// WebSocket against a live server. It proves:
//
//  1. Cancel mid-turn stops it. A `cancel` frame while a turn is parked in a tool
//     cancels the turn's context — the tool's in-flight wait is abandoned and its
//     post-wait line never runs — and a terminal `cancelled` event (status 499) is
//     emitted. No eventual_response follows and the connection stays usable.
//  2. Cancel with no active turn is a silent no-op (no event; connection stays live).
//  3. A normal turn still completes with an eventual_response (the cancellation wiring
//     doesn't disturb the happy path).
//  4. Disconnect mid-turn also aborts the turn (no client remains to receive its output).
//
// Plus the single-active-turn rule the same spec mandates (Go-side coverage of what the
// Rust reader loop enforces): a second send_message while a turn is in flight is
// rejected with TURN_IN_PROGRESS rather than run concurrently.
//
// Runs fully offline: a MockLlmProvider scripts a call to a tool that parks the turn on
// a context-aware wait, giving a stable in-flight window to cancel in. No gateway.
//
// Cancellation in Go is COOPERATIVE (a context), not preemptive like dropping a Rust
// future — so the parking tool waits on ctx.Done(), which is what any well-behaved Go
// tool does. That is the one deliberate language-level difference from the reference.

const cancelSlowTool = "slow_probe"

// slowToolProbe records what a parked tool observed: that it started, whether its
// context was cancelled out from under it (the positive signal that the turn was
// abandoned mid-wait), and whether it ever reached its post-wait completion.
type slowToolProbe struct {
	started   chan struct{}
	cancelled atomic.Bool
	finished  atomic.Bool
	startOnce atomic.Bool
}

func newSlowToolProbe() *slowToolProbe {
	return &slowToolProbe{started: make(chan struct{})}
}

// tool builds the parking tool: it signals that it started, then waits far longer than
// any test. A cancelled turn cancels its ctx, so it returns early with cancelled=true
// and never sets finished.
func (p *slowToolProbe) tool() core.Tool {
	return core.FuncTool{
		ToolName: cancelSlowTool,
		Desc:     "parks the turn for cancellation tests",
		Params:   map[string]any{"type": "object"},
		Fn: func(ctx context.Context, _ map[string]any) (string, error) {
			if p.startOnce.CompareAndSwap(false, true) {
				close(p.started)
			}
			select {
			case <-ctx.Done():
				p.cancelled.Store(true)
				return "", ctx.Err()
			case <-time.After(time.Hour):
				// Only reached if the turn was NOT cancelled.
				p.finished.Store(true)
				return "done", nil
			}
		},
	}
}

// awaitStart blocks until the turn is provably parked inside the tool.
func (p *slowToolProbe) awaitStart(t *testing.T) {
	t.Helper()
	select {
	case <-p.started:
	case <-time.After(5 * time.Second):
		t.Fatal("turn never parked in the slow tool")
	}
}

// awaitCancelled polls until the parked tool observed its context cancellation.
func (p *slowToolProbe) awaitCancelled(t *testing.T) {
	t.Helper()
	deadline := time.Now().Add(5 * time.Second)
	for time.Now().Before(deadline) {
		if p.cancelled.Load() {
			return
		}
		time.Sleep(10 * time.Millisecond)
	}
	t.Fatal("timed out waiting for the turn's context to be cancelled")
}

// slowToolServer spins up a server whose mock scripts a single call to the parking tool
// (so the turn parks and never returns on its own), then a text reply that a
// non-cancelled turn would settle with.
func slowToolServer(t *testing.T, probe *slowToolProbe) *LocalServer {
	t.Helper()
	mock := core.NewMockLlmProvider()
	mock.PushToolCall("call-1", cancelSlowTool, `{}`)
	mock.PushText("Finished the slow thing.")

	ls, err := SpawnLocal(
		WithLocalAddr("127.0.0.1:0"),
		WithLocalChatClient(mock),
		WithLocalServerOption(WithTools([]core.Tool{probe.tool()})),
	)
	if err != nil {
		t.Fatalf("spawn: %v", err)
	}
	t.Cleanup(func() { _ = ls.Shutdown() })
	return ls
}

// nextRawEv reads the next server event WITHOUT the pong/keepalive filtering nextEv
// applies — the no-op test asserts that a pong is literally the next thing on the wire.
func nextRawEv(t *testing.T, transport protocol.Transport, within time.Duration) map[string]any {
	t.Helper()
	select {
	case data, ok := <-transport.Receive():
		if !ok {
			t.Fatalf("transport closed before expected event")
		}
		var ev map[string]any
		if err := json.Unmarshal(data, &ev); err != nil {
			t.Fatalf("decode event: %v (raw=%s)", err, data)
		}
		return ev
	case <-time.After(within):
		t.Fatalf("timed out waiting for next event")
		return nil
	}
}

// recvWithin returns the next event, or nil if none arrives within the window (used to
// assert that NOTHING follows a cancellation).
func recvWithin(t *testing.T, transport protocol.Transport, within time.Duration) map[string]any {
	t.Helper()
	select {
	case data, ok := <-transport.Receive():
		if !ok {
			return nil // transport closed — no event
		}
		var ev map[string]any
		if err := json.Unmarshal(data, &ev); err != nil {
			t.Fatalf("decode event: %v (raw=%s)", err, data)
		}
		return ev
	case <-time.After(within):
		return nil
	}
}

// recvUntil reads events until one of type typ arrives, collecting the ones it skipped
// (the Go analog of the Rust common::recv_until).
func recvUntil(t *testing.T, transport protocol.Transport, typ string, within time.Duration) (map[string]any, []map[string]any) {
	t.Helper()
	var seen []map[string]any
	deadline := time.After(within)
	for {
		select {
		case data, ok := <-transport.Receive():
			if !ok {
				t.Fatalf("transport closed waiting for %q (seen=%d events)", typ, len(seen))
			}
			var ev map[string]any
			if err := json.Unmarshal(data, &ev); err != nil {
				t.Fatalf("decode event: %v (raw=%s)", err, data)
			}
			if got, _ := ev["type"].(string); got == typ {
				return ev, seen
			}
			seen = append(seen, ev)
		case <-deadline:
			t.Fatalf("timed out waiting for %q (seen=%d events)", typ, len(seen))
			return nil, nil
		}
	}
}

// TestCancelMidTurnAbortsAndEmitsCancelled — a cancel while the turn is parked in a tool
// aborts the turn and emits the terminal `cancelled` event; nothing follows it, and the
// connection stays usable.
func TestCancelMidTurnAbortsAndEmitsCancelled(t *testing.T) {
	probe := newSlowToolProbe()
	ls := slowToolServer(t, probe)
	transport := connectTransport(t, ls)
	defer func() { _ = transport.Close() }()

	sessionID := createSession(t, transport)

	// Start a turn; it parks inside the slow tool.
	sendFrame(t, transport, map[string]any{
		"action":    "send_message",
		"requestId": "turn-1",
		"sessionId": sessionID,
		"message":   "please do the slow thing",
	})
	probe.awaitStart(t)
	if probe.finished.Load() {
		t.Fatal("tool must not have finished yet")
	}

	// Cancel it, reusing the turn's requestId (the correlation convention).
	sendFrame(t, transport, map[string]any{"action": "cancel", "requestId": "turn-1"})

	// A terminal `cancelled` event arrives, echoing the turn's requestId. (Skip the ack /
	// stream events that were already in flight when the cancel landed.)
	ev, _ := recvUntil(t, transport, "cancelled", 5*time.Second)
	if rid, _ := ev["requestId"].(string); rid != "turn-1" {
		t.Fatalf("cancelled requestId = %v, want turn-1 (event=%s)", ev["requestId"], mustJSON(ev))
	}
	if status, _ := asInt(ev["status"]); status != 499 {
		t.Fatalf("cancelled status = %v, want 499 (event=%s)", ev["status"], mustJSON(ev))
	}
	if rid, ok := dot(t, ev, "data.requestId"); !ok || rid != "turn-1" {
		t.Fatalf("cancelled data.requestId = %v, want turn-1 (event=%s)", rid, mustJSON(ev))
	}
	if status, ok := dot(t, ev, "data.status"); !ok {
		t.Fatalf("cancelled data.status missing (event=%s)", mustJSON(ev))
	} else if n, _ := asInt(status); n != 499 {
		t.Fatalf("cancelled data.status = %v, want 499 (event=%s)", status, mustJSON(ev))
	}
	// No answer payload: a cancelled turn produced no assistant message.
	if _, ok := dot(t, ev, "data.messageId"); ok {
		t.Fatalf("cancelled must carry no messageId (event=%s)", mustJSON(ev))
	}

	// The turn was abandoned mid-wait: the tool saw its context cancelled and never
	// reached its post-wait completion.
	probe.awaitCancelled(t)
	if probe.finished.Load() {
		t.Fatal("cancelled turn's tool must never reach its post-wait completion")
	}

	// No further terminal event (no eventual_response) follows the cancellation.
	if after := recvWithin(t, transport, 500*time.Millisecond); after != nil {
		t.Fatalf("no event should follow the cancellation, got: %s", mustJSON(after))
	}

	// The connection is still alive and usable.
	sendFrame(t, transport, map[string]any{"action": "ping", "requestId": "p1"})
	pong := nextRawEv(t, transport, 5*time.Second)
	if typ, _ := pong["type"].(string); typ != "pong" {
		t.Fatalf("expected pong after cancellation, got %s", mustJSON(pong))
	}
	if rid, _ := pong["requestId"].(string); rid != "p1" {
		t.Fatalf("pong requestId = %v, want p1", pong["requestId"])
	}
}

// TestCancelWithNoActiveTurnIsNoop — a cancel with nothing running emits nothing and
// leaves the connection live.
func TestCancelWithNoActiveTurnIsNoop(t *testing.T) {
	mock := core.NewMockLlmProvider()
	mock.PushText("hi")
	ls, err := SpawnLocal(WithLocalAddr("127.0.0.1:0"), WithLocalChatClient(mock))
	if err != nil {
		t.Fatalf("spawn: %v", err)
	}
	t.Cleanup(func() { _ = ls.Shutdown() })

	transport := connectTransport(t, ls)
	defer func() { _ = transport.Close() }()
	_ = createSession(t, transport)

	// Cancel with nothing running: must emit nothing.
	sendFrame(t, transport, map[string]any{"action": "cancel", "requestId": "nope"})

	// So the NEXT event on the wire is the pong (the cancel produced no event of its own).
	sendFrame(t, transport, map[string]any{"action": "ping", "requestId": "p1"})
	ev := nextRawEv(t, transport, 5*time.Second)
	if typ, _ := ev["type"].(string); typ != "pong" {
		t.Fatalf("cancel must not emit an event; got: %s", mustJSON(ev))
	}
	if rid, _ := ev["requestId"].(string); rid != "p1" {
		t.Fatalf("pong requestId = %v, want p1", ev["requestId"])
	}
}

// TestNormalTurnStillCompletes — the cancellation wiring doesn't disturb the happy path.
func TestNormalTurnStillCompletes(t *testing.T) {
	mock := core.NewMockLlmProvider()
	mock.PushText("All done here.")
	ls, err := SpawnLocal(WithLocalAddr("127.0.0.1:0"), WithLocalChatClient(mock))
	if err != nil {
		t.Fatalf("spawn: %v", err)
	}
	t.Cleanup(func() { _ = ls.Shutdown() })

	transport := connectTransport(t, ls)
	defer func() { _ = transport.Close() }()
	sessionID := createSession(t, transport)

	sendFrame(t, transport, map[string]any{
		"action":    "send_message",
		"requestId": "turn-ok",
		"sessionId": sessionID,
		"message":   "hello",
	})

	done, seen := recvUntil(t, transport, "eventual_response", 10*time.Second)
	if rid, _ := done["requestId"].(string); rid != "turn-ok" {
		t.Fatalf("eventual_response requestId = %v, want turn-ok (event=%s)", done["requestId"], mustJSON(done))
	}
	if status, _ := asInt(done["status"]); status != 200 {
		t.Fatalf("eventual_response status = %v, want 200", done["status"])
	}
	for _, ev := range seen {
		if typ, _ := ev["type"].(string); typ == "cancelled" {
			t.Fatalf("a normal turn must not emit a cancelled event: %s", mustJSON(ev))
		}
	}
}

// TestDisconnectMidTurnAbortsTheTurn — the client hanging up mid-turn aborts the turn
// (no client remains to receive its output).
func TestDisconnectMidTurnAbortsTheTurn(t *testing.T) {
	probe := newSlowToolProbe()
	ls := slowToolServer(t, probe)
	transport := connectTransport(t, ls)

	sessionID := createSession(t, transport)
	sendFrame(t, transport, map[string]any{
		"action":    "send_message",
		"requestId": "turn-x",
		"sessionId": sessionID,
		"message":   "please do the slow thing",
	})
	probe.awaitStart(t)

	// Client hangs up mid-turn.
	if err := transport.Close(); err != nil {
		t.Fatalf("close transport: %v", err)
	}

	// The server aborts the in-flight turn: the tool's context is cancelled and it never
	// reaches its post-wait completion.
	probe.awaitCancelled(t)
	if probe.finished.Load() {
		t.Fatal("disconnect must abort the turn before it completes")
	}
}

// TestSecondSendMessageWhileTurnInFlightIsRejected — ONE active turn per connection: a
// second send_message while one is running is rejected with TURN_IN_PROGRESS, never run
// concurrently.
func TestSecondSendMessageWhileTurnInFlightIsRejected(t *testing.T) {
	probe := newSlowToolProbe()
	ls := slowToolServer(t, probe)
	transport := connectTransport(t, ls)
	defer func() { _ = transport.Close() }()

	sessionID := createSession(t, transport)
	sendFrame(t, transport, map[string]any{
		"action":    "send_message",
		"requestId": "turn-1",
		"sessionId": sessionID,
		"message":   "please do the slow thing",
	})
	probe.awaitStart(t)

	// A second turn on the same socket while the first is parked.
	sendFrame(t, transport, map[string]any{
		"action":    "send_message",
		"requestId": "turn-2",
		"sessionId": sessionID,
		"message":   "and another thing",
	})

	ev, _ := recvUntil(t, transport, "error", 5*time.Second)
	if rid, _ := ev["requestId"].(string); rid != "turn-2" {
		t.Fatalf("error requestId = %v, want turn-2 (event=%s)", ev["requestId"], mustJSON(ev))
	}
	code, ok := dot(t, ev, "error.code")
	if !ok || code != "TURN_IN_PROGRESS" {
		t.Fatalf("error code = %v, want TURN_IN_PROGRESS (event=%s)", code, mustJSON(ev))
	}

	// The rejected frame started no turn: cancelling clears the ONE turn that is running,
	// and the connection settles with a single `cancelled`.
	sendFrame(t, transport, map[string]any{"action": "cancel", "requestId": "turn-1"})
	got, _ := recvUntil(t, transport, "cancelled", 5*time.Second)
	if rid, _ := got["requestId"].(string); rid != "turn-1" {
		t.Fatalf("cancelled requestId = %v, want turn-1 (event=%s)", got["requestId"], mustJSON(got))
	}
}

File: typescript/server/test/turn-cancel.test.ts
/**
 * User-initiated turn cancellation — the `cancel` action ("Stop button").
 *
 * The TS parity port of the Rust reference's `rust/smooth-operator-server/tests/turn_cancel.rs`.
 * It proves, over a real WebSocket:
 *
 *   1. **Cancel mid-turn stops it.** A `cancel` frame while a turn is parked in a tool
 *      aborts the turn — a terminal `cancelled` event (status 499) is emitted, no
 *      `eventual_response` follows, the model is never called again, and the partial
 *      assistant reply is DISCARDED while the user's message stays persisted.
 *   2. **Cancel with no active turn is a silent no-op** (no event; connection stays live).
 *   3. **A normal turn still completes** with an `eventual_response` (the cancellation
 *      wiring doesn't disturb the happy path).
 *   4. **Disconnect mid-turn also aborts the turn** (no client remains to receive it).
 *
 * Plus the single-active-turn rule the reference enforces alongside cancel: a second
 * `send_message` while one is in flight is rejected with `TURN_IN_PROGRESS`.
 *
 * Runs fully offline: a `MockLlmProvider` scripts the turn and a host tool parks it on a
 * gate the test releases, giving a stable in-flight window to cancel in.
 *
 * Where this differs from Rust: JS cannot DROP an in-flight `await` the way tokio's
 * `abort()` drops a future, so the reference's drop-guard assertion has no TS analogue.
 * Cancellation is cooperative — what is asserted instead is the observable protocol
 * contract (terminal `cancelled`, no `eventual_response`, no further model call, no
 * persisted reply), which is the part every language server must match.
 */
import { MockLlmProvider } from '@smooai/smooth-operator-core';
import type { Tool } from '@smooai/smooth-operator-core';
import { afterEach, describe, expect, it } from 'vitest';

import { serve, type RunningServer } from '../src/server.js';
import { InMemorySessionStore } from '../src/sessionStore.js';
import { TestClient } from './wsClient.js';

const SLOW_TOOL = 'slow_probe';

/**
 * A tool that PARKS the turn: it records that it started, then blocks on a gate the test
 * controls. The TS stand-in for the Rust test's hour-long `tokio::time::sleep` — with an
 * explicit release so no timer outlives the test.
 */
class SlowTool {
    started = false;
    finished = false;
    private release!: () => void;
    private readonly gate = new Promise<void>((resolve) => {
        this.release = resolve;
    });

    /** Let the parked tool return. */
    finish(): void {
        this.release();
    }

    tool(): Tool {
        return {
            name: SLOW_TOOL,
            description: 'parks the turn for cancellation tests',
            parameters: { type: 'object', properties: {} },
            execute: async (): Promise<string> => {
                this.started = true;
                await this.gate;
                this.finished = true;
                return 'done';
            },
        };
    }
}

/** A mock that calls the slow tool (so the turn parks in it), then would wrap up. */
function slowToolMock(): MockLlmProvider {
    const mock = new MockLlmProvider();
    mock.pushToolCall('call_1', SLOW_TOOL, '{}');
    mock.pushText('All done here.');
    return mock;
}

/** A mock that just answers (no tool), so a turn completes normally. */
function answerMock(text: string): MockLlmProvider {
    const mock = new MockLlmProvider();
    mock.pushText(text);
    return mock;
}

/** Drive `create_conversation_session` and return the session + conversation ids. */
async function createSession(client: TestClient): Promise<{ sessionId: string; conversationId: string }> {
    client.sendAction({ action: 'create_conversation_session', requestId: 'cs-1', agentId: '11111111-1111-1111-1111-111111111111' });
    const created = await client.receive();
    expect(created.type).toBe('immediate_response');
    const data = created.data as { sessionId: string; conversationId: string };
    return { sessionId: data.sessionId, conversationId: data.conversationId };
}

/** Try to receive one event within `ms`; `undefined` when none arrives (the no-event assertion). */
async function recvWithin(client: TestClient, ms: number): Promise<Record<string, unknown> | undefined> {
    try {
        return await client.receive(ms);
    } catch {
        return undefined;
    }
}

/** Poll `predicate` until it holds or the deadline elapses. */
async function waitFor(predicate: () => boolean, timeoutMs = 5000): Promise<void> {
    const deadline = Date.now() + timeoutMs;
    while (Date.now() < deadline) {
        if (predicate()) return;
        await new Promise((r) => setTimeout(r, 10));
    }
    throw new Error('condition not met within timeout');
}

describe('turn cancellation — the cancel action', () => {
    let server: RunningServer | undefined;
    let slow: SlowTool | undefined;

    afterEach(async () => {
        // Release any parked tool first so teardown never waits on the gate.
        slow?.finish();
        slow = undefined;
        await server?.close();
        server = undefined;
    });

    it('cancel mid-turn aborts the turn and emits a terminal cancelled event', async () => {
        slow = new SlowTool();
        const mock = slowToolMock();
        const store = new InMemorySessionStore();
        server = await serve({ chatClient: mock, store, tools: [slow.tool()] });

        const client = await TestClient.connect(server.url);
        try {
            const { sessionId, conversationId } = await createSession(client);
            client.sendAction({ action: 'send_message', requestId: 'turn-1', sessionId, message: 'please do the slow thing' });

            const ack = await client.receive();
            expect(ack.status).toBe(202);

            // Wait until the turn is genuinely in flight (parked inside the tool).
            await waitFor(() => slow!.started);
            expect(slow.finished).toBe(false);

            // Cancel it, reusing the turn's requestId (the correlation convention).
            client.sendAction({ action: 'cancel', requestId: 'turn-1' });

            const { terminal } = await client.receiveUntil('cancelled');
            expect(terminal.requestId).toBe('turn-1');
            expect(terminal.status).toBe(499);
            expect(terminal.data).toEqual({ requestId: 'turn-1', status: 499 });

            // Let the parked tool return: the turn observes the cancel and stops there.
            slow.finish();

            // No further terminal event (no eventual_response) follows the cancellation…
            const after = await recvWithin(client, 500);
            expect(after).toBeUndefined();
            // …and the turn never continued: the model was never called a second time.
            expect(mock.callCount).toBe(1);

            // Partial output discarded: the user's message stays persisted, the assistant
            // reply never is.
            const messages = await store.listMessages(conversationId, 50);
            expect(messages.map((m) => m.direction)).toEqual(['inbound']);

            // The connection is still alive and usable.
            client.sendAction({ action: 'ping', requestId: 'p1' });
            const pong = await client.receive();
            expect(pong.type).toBe('pong');
            expect(pong.requestId).toBe('p1');
        } finally {
            await client.close();
        }
    });

    it('cancel with no active turn is a silent no-op', async () => {
        server = await serve({ chatClient: answerMock('hi') });
        const client = await TestClient.connect(server.url);
        try {
            await createSession(client);

            client.sendAction({ action: 'cancel', requestId: 'nope' });

            // The next event is the pong — the cancel produced no event of its own.
            client.sendAction({ action: 'ping', requestId: 'p1' });
            const event = await client.receive();
            expect(event.type).toBe('pong');
            expect(event.requestId).toBe('p1');
        } finally {
            await client.close();
        }
    });

    it('a normal turn still completes with an eventual_response', async () => {
        server = await serve({ chatClient: answerMock('All done here.') });
        const client = await TestClient.connect(server.url);
        try {
            const { sessionId } = await createSession(client);
            client.sendAction({ action: 'send_message', requestId: 'turn-ok', sessionId, message: 'hello' });

            const { terminal, seen } = await client.receiveUntil('eventual_response');
            expect(terminal.requestId).toBe('turn-ok');
            expect(terminal.status).toBe(200);
            expect(seen.some((e) => e.type === 'cancelled')).toBe(false);
        } finally {
            await client.close();
        }
    });

    it('disconnect mid-turn aborts the in-flight turn', async () => {
        slow = new SlowTool();
        const mock = slowToolMock();
        const store = new InMemorySessionStore();
        server = await serve({ chatClient: mock, store, tools: [slow.tool()] });

        const client = await TestClient.connect(server.url);
        const { sessionId, conversationId } = await createSession(client);
        client.sendAction({ action: 'send_message', requestId: 'turn-x', sessionId, message: 'please do the slow thing' });
        expect((await client.receive()).status).toBe(202);
        await waitFor(() => slow!.started);

        // Client hangs up mid-turn. `close()` resolves on the CLIENT's socket; give the
        // server a beat to observe the close and abort the turn before releasing the tool.
        await client.close();
        await new Promise((r) => setTimeout(r, 200));

        // Release the parked tool; the server must abort rather than finish the turn.
        slow.finish();
        await new Promise((r) => setTimeout(r, 300));

        expect(mock.callCount).toBe(1);
        const messages = await store.listMessages(conversationId, 50);
        expect(messages.map((m) => m.direction)).toEqual(['inbound']);
    });

    it('a second send_message while a turn is in flight is rejected with TURN_IN_PROGRESS', async () => {
        slow = new SlowTool();
        const store = new InMemorySessionStore();
        server = await serve({ chatClient: slowToolMock(), store, tools: [slow.tool()] });

        const client = await TestClient.connect(server.url);
        try {
            const { sessionId } = await createSession(client);
            client.sendAction({ action: 'send_message', requestId: 'turn-1', sessionId, message: 'first' });
            expect((await client.receive()).status).toBe(202);
            await waitFor(() => slow!.started);

            client.sendAction({ action: 'send_message', requestId: 'turn-2', sessionId, message: 'second' });
            // Skip the first turn's in-flight stream frames (its toolCall chunk).
            const { terminal: err } = await client.receiveUntil('error');
            expect(err.requestId).toBe('turn-2');
            expect((err.error as { code: string }).code).toBe('TURN_IN_PROGRESS');

            // Cancelling the first frees the connection for a new turn.
            client.sendAction({ action: 'cancel', requestId: 'turn-1' });
            const { terminal } = await client.receiveUntil('cancelled');
            expect(terminal.requestId).toBe('turn-1');
            slow.finish();
        } finally {
            await client.close();
        }
    });
});
